### ISSUE TITLE
Race condition on shutdown can cause SEGV

### ISSUE THREAD

**Issue by Geod24** [bosagora/agora#2942]
The small commit introduced a regression:
```
Unittesting agora.test.Flash#### FATAL ERROR: index [0] exceeds array of length 0
This node was started at source/agora/test/Flash.d:914
This most likely means that the node crashed due to an uncaught exception
If not, please file a bug at https://github.com/Geod24/localrest/
Full error: core.exception.ArrayIndexError@source/agora/node/FullNode.d(1159): index [0] exceeds array of length 0
----------------
core/exception.d:875 onArrayIndexError [0x11132fc6c]
core/exception.d:848 _d_arraybounds_index [0x1113301b5]
source/agora/node/FullNode.d:1159 nothrow @safe void agora.node.FullNode.FullNode.pushBlock(const(agora.consensus.data.Block.Block)).__lambda7() [0x110665b09]
core/thread/context.d:46 void core.thread.context.Callable.opCall() [0x11135c34c]
core/thread/fiber.d:1151 void core.thread.fiber.Fiber.run() [0x11135c4d8]
core/thread/fiber.d:198 fiber_entryPoint [0x11135c415]
./ci/run.sh: line 21:  3795 Abort trap: 6           ./build/agora-unittests
```

_Originally posted by @Geod24 in https://github.com/bosagora/agora/issues/2915#issuecomment-1024529269_

This is because the task is still running (but shouldn't).

**Comment by Geod24** [comment]
We probably need a way to shut down the task manager. We might even want to, IDK, name tasks ?

**Comment by hewison-chris** [comment]
I think this was fixed by 
https://github.com/bosagora/agora/commit/e86c0d06d1761bd4470628cf85d4d2d7052b3e63
and
https://github.com/bosagora/agora/commit/093103ba12d25ef0cc85056a07bfb3d1951bcdb1

**Comment by hewison-chris** [comment]
However this might still be related https://github.com/bosagora/agora/issues/3079

**Comment by linked0** [comment]
This could be solved by PR #3109.

**Comment by linked0** [comment]
This is probably solved by [PR #3109](https://github.com/bosagora/agora/pull/3109). So closing this.

**Comment by linked0** [comment]
I see the issue happens again. So reopening this.
